### ISSUE TITLE
Inclusion with Odin's `core` library

### ISSUE THREAD

**Issue by gingerBill** [Feoramund/odin-varg#1]
Would you be interested in adding this into Odin's `core` library collection?

-----------

The only "issue" I have with this library is that it only works for Odin-style flags:
```
-flag
-flag:key
-flag:key=value
```

I made Odin's style that way for three reasons: 

* Each argument can be parsed independently from the `argv`/`os.args`
* It's a context-free grammar
* It's extremely consistent and simple

However, the "issue" I bring up is that certain tooling wants a more GNU-like style
```
--flag
--flag=argument
--flag argument
--flag argument repeating-argument
-f         (short flags)
-fghi      (combining flags)
```

Personally I don't like the short flags nor combination of short flags, as they are usually cryptic to me. The other reason this style might be preferred is for the arguments separated by spaces being produced from another program. This is common enough in the UNIX world that it might need to be done.

**TO CLARIFY**

I would LOVE this in Odin's `core` even with just Odin-style flags.

**Comment by Feoramund** [comment]
> Would you be interested in adding this into Odin's `core` library collection?

Absolutely.

> I made Odin's style that way for three reasons:
> 
> * Each argument can be parsed independently from the `argv`/`os.args`
> * It's a context-free grammar
> * It's extremely consistent and simple
>
> [...]
>
> Personally I don't like the short flags nor combination of short flags, as they are usually cryptic to me.

You're not alone in that. I always have to go digging through documentation to remember what they're for. This is why I made the library as it is; I just found Odin's style so much more consistent and understandable.

> The other reason this style might be preferred is for the arguments separated by spaces being produced from another program. This is common enough in the UNIX world that it might need to be done. 

With the idea of tooling in mind and from a design standpoint, I think supporting spaced arguments might be alright.
```
-flag argument
-flag argument repeating-argument
```
I'll have to dig into the code again and see how tough that'd be to get working though. The first issue I can see is that a variadic flag with spaces would have to come last, as it'd eat up all the positionals otherwise, and I don't know how you disambiguate a case where you have a variadic/dynamic array `flag` and `-flag a b c -flag2 d`. Is `-flag2` an argument? A flag? Who knows... I try to avoid writing disambiguation logic.

I'll need to think about how to best handle this style, since the parser isn't made to rely on prior information.

I do however agree to the notion that encouraging a bit of friction is good where the outcome produces better results. I don't want short flags, flag aliases, or short combining flags for the reasons we both dislike them. If we don't support these sort of things, hopefully the programs people make won't be as cryptic as the ones we've seen.

> **TO CLARIFY**
> 
> I would LOVE this in Odin's `core` even with just Odin-style flags.

Thank you for taking the time to review the library and reach out.

Even if I don't make any changes to support extra argument styles, I'll have to give the library a once-over when I have some time to make sure I'm satisfied with sending it in as a proper PR to Odin. I know the RTTI-handling stuff was a bit crunchy with all the switch statements. I remembering thinking it could be simplified, but I'll have to peek at it again.
